### DISCUSSION TITLE
Missing sysinfo with per test config

### DISCUSSION THREAD

**Discussion by pevogam** [avocado-framework/avocado#5681] [Q&A] [answered]
Hi all, I am using the following config option in a default avocado config file in order to enable the now available sysinfo per test:
```
[sysinfo.collect]
# Whether to collect sysinfo per test or just once for the entire job
per_test = True
```
However I find no respective sysinfo folders within the test results folder. Could it be that I am missing something obvious here?

**Comment by richtja** [comment]
Hi @pevogam, if you have sysinfo per job enabled before, then the `sysinfo.collect.per_test = True` should be the only thing you have to do. Can you please check if you have  `sysinfo.collect.enabled = True` (default value)  and also that the sysinfo collectibles points to the right paths on your system:
``` 
sysinfo.collectibles.commands
sysinfo.collectibles.fail_commands  
sysinfo.collectibles.fail_files
sysinfo.collectibles.files 
sysinfo.collectibles.profilers 
```
If all of this is correct, then you should see sysinfo collection per test in `$RESULTS/test-results/$TEST/sysinfo`. If you still don't get the sysinfo directory, we will have to investigate it, it might be a bug.

**Reply by pevogam** [reply]
This all the configuration that takes effect, the config is respected in that regard:
```
[root@c5 intra2net-qa]# avocado config | grep sysinfo
    diff.filter                              ['cmdline', 'time', 'variants', 'results', 'config', 'sysinfo']
    sysinfo.collect.commands_timeout         -1
    sysinfo.collect.enabled                  True
    sysinfo.collect.installed_packages       False
    sysinfo.collect.locale                   C
    sysinfo.collect.optimize                 False
    sysinfo.collect.per_test                 True
    sysinfo.collect.profiler                 False
    sysinfo.collect.sysinfodir               None
    sysinfo.collectibles.commands            /etc/avocado/sysinfo/commands
    sysinfo.collectibles.fail_commands       /usr/lib/python3.10/site-packages/avocado/etc/avocado/sysinfo/fail_commands
    sysinfo.collectibles.fail_files          /usr/lib/python3.10/site-packages/avocado/etc/avocado/sysinfo/fail_files
    sysinfo.collectibles.files               /etc/avocado/sysinfo/files
    sysinfo.collectibles.profilers           /etc/avocado/sysinfo/profilers
```
I managed to also reproduce part of this with a vanilla VT test:
```
[root@c5 avocado-i2n-98.0]# avocado run --status-server-disable-auto --spawner=lxc boot
/usr/lib/python3.10/site-packages/avocado_vt/plugins/vt_resolver.py:47: UserWarning: The VT NextRunner can be run only with max-parallel-tasks set to 1
  warnings.warn("The VT NextRunner can be run only "
JOB ID     : 1aa842565703d5f49364f25eb5c9f0183f405ce2
JOB LOG    : /mnt/local/results/job-2023-05-16T19.30-1aa8425/job.log
 (1/1) io-github-autotest-qemu.boot: STARTED
 (1/1) io-github-autotest-qemu.boot: PASS (28.21 s)
RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /mnt/local/results/job-2023-05-16T19.30-1aa8425/results.html
JOB TIME   : 33.48 s
[root@c5 avocado-i2n-98.0]# ls /mnt/local/results/job-2023-05-16T19.30-1aa8425/test-results/1-io-github-autotest-qemu.boot/sysinfo/
pre
```
where the post folder and all its files are missing. The complete sysinfo missing then seems to be another thing related to our use of custom scheduler. Would you mind pointing me to the location in Avocado VT that had to be adapted to have the `pre` folder and has to be adapted to have the `post` folder available?

**Reply by richtja** [reply]
I was able to reproduce it and I found a bug there. You can see fix in #5682. Thank you for catching this.

**Reply by pevogam** [reply]
That's great, I will take a look. However, it seems my concern would go beyond this - The complete sysinfo is still missing with our custom scheduler which brings me to two conclusions:
1) There are missing lines or something we have to do per test so that the sysinfo parameter is respected. Could you point me to something that avocado-vt or the default state machine does to set this up?
2) More importantly than the previous point, as a usual plugin developer for avocado I would expect that a custom scheduler is not supposed to be affected by anything missing and the sysinfo should be decoupled from it. This means that either on our side we are touching doing something we are not supposed to or on avocado's side there are not clear boundaries on what a plugin developer should expect to be provided with and must integrate with in return. I guess the way to figure this larger point is to look at details from (1) but ultimately I wonder if we can establish a clear boundary for what an alternative scheduler plugin could do and should not cross into.

**Reply by richtja** [reply]
> That's great, I will take a look. However, it seems my concern would go beyond this - The complete sysinfo is still missing with our custom scheduler which brings me to two conclusions:
> 
>    1. There are missing lines or something we have to do per test so that the sysinfo parameter is respected. Could you point me to something that avocado-vt or the default state machine does to set this up?
> 
On avocado-vt side doesn't have to do anything to set up sysinfo per tests. The necessary dependencies to make this work are post test plugins' mechanism introduced in #5630 and Runnable log_dir changes introduced in  #5660. Those are huge changes, but unfortunately without knowledge of your code I can't point you into a more specific place. I hope that this will help, if not we can try to investigate more deeply together.

>    2. More importantly than the previous point, as a usual plugin developer for avocado I would expect that a custom scheduler is not supposed to be affected by anything missing and the sysinfo should be decoupled from it. This means that either on our side we are touching doing something we are not supposed to or on avocado's side there are not clear boundaries on what a plugin developer should expect to be provided with and must integrate with in return. I guess the way to figure this larger point is to look at details from (1) but ultimately I wonder if we can establish a clear boundary for what an alternative scheduler plugin could do and should not cross into.

I agree with you that there have to be clear boundaries and we are trying to do it by specific namespaces. We have avocado.core for avocado internal and avocado.plugins for plugins which has pre-defined interface which shouldn't change (might be extended, but with back port support). I know that we have problem with avocado.plugins.runner_nrunner plugin which is a runner for suites. Even that this is a plugin, it is closely connected to core and this connection might create troubles for projects which has its own SuiteRunner. And I agree we should discuss how to make boundaries here.

I am not sure what you mean by custom scheduler. If you have your own avocado.core.task.statemachine than it is part of avocado core and it is possible that avocado changes might cause trouble to you, but we do not support it. If you meant avocado.plugins.runner_nrunner plugin, then as I said that is something we have to discuss and fix. Maybe we just need to add tests which will protect avocado.plugins.runner_nrunner plugin interface.

**Reply by pevogam** [reply]
> > That's great, I will take a look. However, it seems my concern would go beyond this - The complete sysinfo is still missing with our custom scheduler which brings me to two conclusions:
> > 
> > 1. There are missing lines or something we have to do per test so that the sysinfo parameter is respected. Could you point me to something that avocado-vt or the default state machine does to set this up?
> 
> On avocado-vt side doesn't have to do anything to set up sysinfo per tests. The necessary dependencies to make this work are post test plugins' mechanism introduced in #5630 and Runnable log_dir changes introduced in #5660. Those are huge changes, but unfortunately without knowledge of your code I can't point you into a more specific place. I hope that this will help, if not we can try to investigate more deeply together.

I looked through the changes including separately at each commit in order to understand how this mechanism works internally (so far unfortunately needed to find the problem on our side) and how the sysinfo example was added. IIUC somewhere in the code new tasks are created and added as a pre- and post-test stage to collect the system info before and after the test runs. I failed to see where these tasks are created since the first set of changes does not enable sysinfo and mostly defines the plugins, entry points, and configuraiton (which should all work out of the box for an external plugin) while the second one enables sysinfo but doesn't do so in any way that would involve the state machine or internal avocado.core changes. Could you point me to a precise location where these tasks are created and I apologize if I have missed them from the changes. Also let me try to do some of the reverse pointers too - the previous limitations of the state machine (from some years ago mostly) and in particular discussion like [this one](https://github.com/avocado-framework/avocado/discussions/4617) meant I had to call avocado in a serial manner and have better control over the test traversal by calling each test like so:
```python
        # TODO: use a single state machine for all test nodes when we are able
        # to at least add requested tasks to it safely (using its locks)
        await Worker(state_machine=TaskStateMachine([task], self.status_repo),
                     spawner=node.spawner, max_running=1,
                     task_timeout=job.config.get('task.timeout.running')).run()
```
I assume I have to create a post-task for the purpose or something like this here or find a standard mechanism to read out and queue in all pre- and post- test tasks, one test at a time. I did notice that one of your changes included something very similar to what I mentioned in the discussion above which is a great progress in making the state machine class easier to use and more flexible.

> > 2. More importantly than the previous point, as a usual plugin developer for avocado I would expect that a custom scheduler is not supposed to be affected by anything missing and the sysinfo should be decoupled from it. This means that either on our side we are touching doing something we are not supposed to or on avocado's side there are not clear boundaries on what a plugin developer should expect to be provided with and must integrate with in return. I guess the way to figure this larger point is to look at details from (1) but ultimately I wonder if we can establish a clear boundary for what an alternative scheduler plugin could do and should not cross into.
> 
> I agree with you that there have to be clear boundaries and we are trying to do it by specific namespaces. We have avocado.core for avocado internal and avocado.plugins for plugins which has pre-defined interface which shouldn't change (might be extended, but with back port support). I know that we have problem with avocado.plugins.runner_nrunner plugin which is a runner for suites. Even that this is a plugin, it is closely connected to core and this connection might create troubles for projects which has its own SuiteRunner. And I agree we should discuss how to make boundaries here.
> 
> I am not sure what you mean by custom scheduler. If you have your own avocado.core.task.statemachine than it is part of avocado core and it is possible that avocado changes might cause trouble to you, but we do not support it. If you meant avocado.plugins.runner_nrunner plugin, then as I said that is something we have to discuss and fix. Maybe we just need to add tests which will protect avocado.plugins.runner_nrunner plugin interface.

As you can see in the example above I am generally not touching the state machine in any way and I try to stay away from internal usage (where there are also frequent changes and possibilities of unforeseen breakage) as much as possible. The way I approach the traversal is to have
```python
from avocado.core.plugin_interfaces import SuiteRunner as RunnerInterface
...
class CartesianRunner(RunnerInterface):
```
which has expected runner interface (now named `SuiteRunner` but I needed backward compatibility) methods like `run_suite` and so on. More details about the runner class and the way it wraps around the state machine and inherits the runner interface can be seen [here](https://github.com/intra2net/avocado-i2n/blob/9fe4e2a688adae71b1ddccc0ba7b6a8caaaf6e0e/avocado_i2n/runner.py).

So if we have to advance a bit in direction of clearing such boundaries I would say that defining what to expect and what not to expect from the `SuiteRunner` is a good start.

**Reply by richtja** [reply]
> > > That's great, I will take a look. However, it seems my concern would go beyond this - The complete sysinfo is still missing with our custom scheduler which brings me to two conclusions:
> > > 
> > > 1. There are missing lines or something we have to do per test so that the sysinfo parameter is respected. Could you point me to something that avocado-vt or the default state machine does to set this up?
> > 
> > 
> > On avocado-vt side doesn't have to do anything to set up sysinfo per tests. The necessary dependencies to make this work are post test plugins' mechanism introduced in #5630 and Runnable log_dir changes introduced in #5660. Those are huge changes, but unfortunately without knowledge of your code I can't point you into a more specific place. I hope that this will help, if not we can try to investigate more deeply together.
> 
> I looked through the changes including separately at each commit in order to understand how this mechanism works internally (so far unfortunately needed to find the problem on our side) and how the sysinfo example was added. IIUC somewhere in the code new tasks are created and added as a pre- and post-test stage to collect the system info before and after the test runs. I failed to see where these tasks are created since the first set of changes does not enable sysinfo and mostly defines the plugins, entry points, and configuraiton (which should all work out of the box for an external plugin) while the second one enables sysinfo but doesn't do so in any way that would involve the state machine or internal avocado.core changes. Could you point me to a precise location where these tasks are created and I apologize if I have missed them from the changes. Also let me try to do some of the reverse pointers too - the previous limitations of the state machine (from some years ago mostly) and in particular discussion like [this one](https://github.com/avocado-framework/avocado/discussions/4617) meant I had to call avocado in a serial manner and have better control over the test traversal by calling each test like so:
> 
> ```python
>         # TODO: use a single state machine for all test nodes when we are able
>         # to at least add requested tasks to it safely (using its locks)
>         await Worker(state_machine=TaskStateMachine([task], self.status_repo),
>                      spawner=node.spawner, max_running=1,
>                      task_timeout=job.config.get('task.timeout.running')).run()
> ```
> 
> I assume I have to create a post-task for the purpose or something like this here or find a standard mechanism to read out and queue in all pre- and post- test tasks, one test at a time. I did notice that one of your changes included something very similar to what I mentioned in the discussion above which is a great progress in making the state machine class easier to use and more flexible.

Thank you for this description. I will try to describe the pre- / post-test task mechanism more deeply. It is all based on [BP004](https://avocado-framework.readthedocs.io/en/latest/blueprints/BP004.html#id1). From [avocado 95.0](https://avocado-framework.readthedocs.io/en/latest/releases/95_0.html?highlight=Dependency%20graph#misc-changes) the RuntimeTasks are defined inside [RuntimeTaskGraph](https://github.com/avocado-framework/avocado/blob/master/avocado/core/task/runtime.py#LL259C7-L259C23) which connects them base on the dependencies.  The `RuntimeTaskGraph` is crucial for pre- / post-task mechanism, because it creates pre- / post-task and holds connections between individual tasks. You can find the creation of graph and pre- / post-test tasks in [here](https://github.com/avocado-framework/avocado/blob/master/avocado/core/task/runtime.py#L295). Basically, for each runnable, the [Pre/PostRuntimeTask.get_tasks_from_test_task](https://github.com/avocado-framework/avocado/blob/master/avocado/core/task/runtime.py#L185)  will call all pre- / post-test plugins to generate runnables for their pre- / post-test tasks and those runnables are used by [Pre/PostRuntimeTask.from_runnable ](https://github.com/avocado-framework/avocado/blob/master/avocado/core/task/runtime.py#L35)method to create runtime tasks. After this creation, all the runtime tasks are connected in dependency graph [here](https://github.com/avocado-framework/avocado/blob/master/avocado/core/task/runtime.py#LL328C29-L328C29). At the end, the dependency graph is used for sorting the tasks for state machine in [nrunner plugin](https://github.com/avocado-framework/avocado/blob/master/avocado/plugins/runner_nrunner.py#L293). I hope the explanation make sense to you, but feel free to ask me any questions to this mechanism.

> 
> > > 2. More importantly than the previous point, as a usual plugin developer for avocado I would expect that a custom scheduler is not supposed to be affected by anything missing and the sysinfo should be decoupled from it. This means that either on our side we are touching doing something we are not supposed to or on avocado's side there are not clear boundaries on what a plugin developer should expect to be provided with and must integrate with in return. I guess the way to figure this larger point is to look at details from (1) but ultimately I wonder if we can establish a clear boundary for what an alternative scheduler plugin could do and should not cross into.
> > 
> > 
> > I agree with you that there have to be clear boundaries and we are trying to do it by specific namespaces. We have avocado.core for avocado internal and avocado.plugins for plugins which has pre-defined interface which shouldn't change (might be extended, but with back port support). I know that we have problem with avocado.plugins.runner_nrunner plugin which is a runner for suites. Even that this is a plugin, it is closely connected to core and this connection might create troubles for projects which has its own SuiteRunner. And I agree we should discuss how to make boundaries here.
> > I am not sure what you mean by custom scheduler. If you have your own avocado.core.task.statemachine than it is part of avocado core and it is possible that avocado changes might cause trouble to you, but we do not support it. If you meant avocado.plugins.runner_nrunner plugin, then as I said that is something we have to discuss and fix. Maybe we just need to add tests which will protect avocado.plugins.runner_nrunner plugin interface.
> 
> As you can see in the example above I am generally not touching the state machine in any way and I try to stay away from internal usage (where there are also frequent changes and possibilities of unforeseen breakage) as much as possible. The way I approach the traversal is to have
> 
> ```python
> from avocado.core.plugin_interfaces import SuiteRunner as RunnerInterface
> ...
> class CartesianRunner(RunnerInterface):
> ```
> 
> which has expected runner interface (now named `SuiteRunner` but I needed backward compatibility) methods like `run_suite` and so on. More details about the runner class and the way it wraps around the state machine and inherits the runner interface can be seen [here](https://github.com/intra2net/avocado-i2n/blob/9fe4e2a688adae71b1ddccc0ba7b6a8caaaf6e0e/avocado_i2n/runner.py).
> 

So IIUIC you are creating every RuntimeTask here and then you are creating new TaskStateMachine for each task and you are not using the RuntimeTaskGraph for generating pre/post tasks. I think you have two options:

1. Use `RuntimeTaskGraph` to generate `RuntimeTask` for your test and all pre- / post-test tasks related to this test. I would prefer this solution because this should avoid compatibility issues in the future.
2. You can write your own solution based on the PreRuntimeTask and PostRuntimeTask classes, but then you will have to take care about compatibility issues if something will change.

> So if we have to advance a bit in direction of clearing such boundaries I would say that defining what to expect and what not to expect from the `SuiteRunner` is a good start.

I agree.

**Reply by pevogam** [reply]
Ok this explanation was more than helpful and I have now restored sysinfo and arbitrary pre-rest and post-test tasks that might be registered via plugins.

This leaves the bigger question about the "scheduler plugin boundaries" to clear out. In particular, the approach of our suite runner is different whereby it will use its own graph and this graph while somewhat similar to the `RuntimeTaskGraph` above is also quite different - it considers tests as dependencies for other tests (say you would like to test a VPN configuration and its variations can be split into multiple alternative steps, each one of which is a valid test in itself), it uses test nodes and test objects that are traced in stateful composition instead of runtime tasks, and so on. What this implies for the relevant conclusions of this thread is that while I am defining a custom suite runner to inherit from the suite runner interface, this runner also has to replace the runtime graph it uses and this runtime graph does not have an interface and is part of the avocado core components (i.e. not assumed to be swappable by plugins). So I think we will have to consider the possibility of a customizable runtime graph here.

I can summarize my observations in a possible tracker issue or a clean discussion with the correct topic if you want. How does that sound?

**Reply by pevogam** [reply]
Moreover thanks to you great feature of adding tasks I was able to reuse the machine like so
```diff
@@ -152,13 +149,13 @@ class CartesianRunner(RunnerInterface):
             elif spawner == "remote":
                 # TODO: perhaps provide a brand new vanilla session for each integration test?
                 task.spawner_handle = node.get_session_to_net()
-        self.tasks += tasks
-
-        # TODO: use a single state machine for all test nodes when we are able
-        # to at least add requested tasks to it safely (using its locks)
-        await Worker(state_machine=TaskStateMachine(tasks, self.status_repo),
-                     spawner=node.spawner, max_running=1,
-                     task_timeout=job.config.get('task.timeout.running')).run()
+            await self.tsm.add_new_task(task)
+        await Worker(
+                state_machine=self.tsm,
+                spawner=node.spawner,
+                max_running=1,
+                task_timeout=job.config.get('task.timeout.running')
+            ).run()
 
     async def run_test_node(self, node):
         """
```
Thanks a lot for this pull request!

**Reply by pevogam** [reply]
> Ok this explanation was more than helpful and I have now restored sysinfo and arbitrary pre-rest and post-test tasks that might be registered via plugins.

But it seems adding the extra pre- and post-test tasks while having all tests passing, slowed down the run by 50% or so making its duration 150% of what it used to be.

> Moreover thanks to you great feature of adding tasks I was able to reuse the machine like so

This then is not even able to complete with passing tests since the asyncio seems to update the status of some 30 second long tests many minutes later, slowing everything down while also encountering other problems.

-> I guess further integration won't be as easy as I have hoped and I will have to investigate more why reusing the same state machine and creating a few supposedly quick pre- and post-test tasks deteriorates the performance so much.

**Comment by pevogam** [comment, marked as answer]
To answer and thus close this discussion, in order for anyone to use sysinfo and other pre- and post-test plugins with a custom scheduler plugin, they will have to construct the necessary runtime pre-tasks and post-tasks by themselves like in the following use case illustration:
```diff
         task = RuntimeTask(raw_task)
-        if spawner == "lxc":
-            task.spawner_handle = host
-        elif spawner == "remote":
-            # TODO: perhaps provide a brand new vanilla session for each integration test?
-            task.spawner_handle = node.get_session_to_net()
-        self.tasks += [task]
+        pre_tasks = PreRuntimeTask.get_tasks_from_test_task(
+            task,
+            1,
+            self.job.test_results_path,
+            None,
+            status_server_uri,
+            self.job.unique_id,
+            self.test_suite.config,
+        )
+        post_tasks = PostRuntimeTask.get_tasks_from_test_task(
+            task,
+            1,
+            self.job.test_results_path,
+            None,
+            status_server_uri,
+            self.job.unique_id,
+            self.test_suite.config,
+        )
+        tasks = [*pre_tasks, task, *post_tasks]
+        for task in tasks:
+            if spawner == "lxc":
+                task.spawner_handle = host
+            elif spawner == "remote":
+                # TODO: perhaps provide a brand new vanilla session for each integration test?
+                task.spawner_handle = node.get_session_to_net()
+        self.tasks += tasks
 
         # TODO: use a single state machine for all test nodes when we are able
         # to at least add requested tasks to it safely (using its locks)
-        await Worker(state_machine=TaskStateMachine([task], self.status_repo),
+        await Worker(state_machine=TaskStateMachine(tasks, self.status_repo),
                      spawner=node.spawner, max_running=1,
                      task_timeout=job.config.get('task.timeout.running')).run()
```

**Reply by pevogam** [reply]
Most of the credits here go to @richtja for explaining the relevant avocado core code.
